src/sentry/remote_config/docs/api.md

```diff
@@ -53,105 +43,231 @@ Retrieve the project's configuration.
+  ```
+
+### Create Configuration [POST]
+
+Create a new configuration. The version field is ignored on create. Version is initialized to 1.
+
+- Request
+
+  ```json
+  {
+    "data": {
+      "environment": "production",
+      "name": "custom-name",
+      "sample_rate": 1.0,
+      "traces_sample_rate": 0.5
+    }
+  }
+  ```
+
+- Response 201
+
+  ```json
+  {
+    "data": {
+      "environment": "production",
+      "id": "0b88ac27a7b444a6baeb312c0493aed5",
+      "name": "custom-name",
+      "sample_rate": 1.0,
+      "traces_sample_rate": 0.5,
+      "version": 1
     }
   }
   ```
 
-**If no existing configuration exists**
+## Configuration [/projects/<organization_id_or_slug>/<project_id_or_slug>/configurations/<configuration_id>/]
+
+### Get Configuration [GET]
+
+Retrieve a configuration.
 
-- Response 404
+- Response 200
 
-### Set Configuration [POST]
+  ```json
+  {
+    "data": {
+      "environment": null,
+      "id": "0b88ac27a7b444a6baeb312c0493aed5",
+      "name": null,
+      "sample_rate": 0,
+      "traces_sample_rate": 0.75,
+      "version": 1
+    }
+  }
+  ```
+
+### Update Configuration [PATCH]
 
-Set the project's configuration.
+Update a configuration. The version attribute is required for all update requests and must match the current server configuration in order to update else an error is returned.
 
 - Request
 
   ```json
   {
     "data": {
-      "features": [
-        {
-          "key": "hello",
-          "value": "world"
-        },
-        {
-          "key": "has_access",
-          "value": true
-        }
-      ],
-      "options": {
-        "sample_rate": 1.0,
-        "traces_sample_rate": 0.5
-      }
+      "sample_rate": 0,
+      "version": 2
     }
   }
   ```
 
-- Response 201
+- Response 202
 
   ```json
   {
     "data": {
-      "features": [
-        {
-          "key": "hello",
-          "value": "world"
-        },
-        {
-          "key": "has_access",
-          "value": true
-        }
-      ],
-      "options": {
-        "sample_rate": 1.0,
-        "traces_sample_rate": 0.5
-      }
+      "environment": "production",
+      "id": "0b88ac27a7b444a6baeb312c0493aed5",
+      "name": null,
+      "sample_rate": 0,
+      "traces_sample_rate": 0.5,
+      "version": 2
     }
   }
   ```
 
 ### Delete Configuration [DELETE]
 
-Delete the project's configuration.
+Delete a configuration.
 
 - Response 204
 
-## Configuration Proxy [/remote-config/projects/<project_id>/]
+## Features [/projects/<organization_id_or_slug>/<project_id_or_slug>/configuration/<configuration_id>/features/]
 
-Temporary configuration proxy resource.
+### Get Features [GET]
 
-### Get Configuration [GET]
+Retrieve configuration features.
 
-Fetch a project's configuration. Responses should be proxied exactly to the SDK.
+- Response 200
+
+  ```json
+  {
+    "data": [
+      {
+        "description": "A feature description.",
+        "is_enabled": true,
+        "key": "hello",
+        "value": 22.3
```

Comment:
> unless we also have a type: 'number' field to tell the SDK and the Settings UI about it

You should have that since the type of the value field would be a number.

> But when the user comes back to edit the form we wouldn't know their preference unless we inspect each value... and that would be annoying if the user wanted to change the value during an edit session...

Ultimately I want the user to have a simple UI experience.  I don't want them to paste JSON or to type the keyword `true` when they want to control if the feature is on or off.  On update if we can inspect the type and provide a simple interface to interact with then great.  On create (and update when changing types) a type-drop-down or something could be used to generate a helpful interface.

> I guess during an edit session the user could just blow away an existing key/value pair then create a new one with a different type and then click save..

I want to avoid this because eventually people will have complicated rule-sets and roll-outs and variants and the rest.  They won't want to lose that data.

Reply:
The SDK knows the type because the user encodes it into their request.  E.g. `sdk.get_boolean_value(key)`

Reply:
> The SDK knows the type because the user encodes it into their request. E.g. sdk.get_boolean_value(key)
IMO, best case is the user has picked a request method (`get_boolean`) and the value needs to be coerced into that type and returned.

But there are issues with that, i think i mentioned type coersion before in a comment, can't find it just now. But we have to be aware of different deployed app versions with different request methods, and current/cached values that don't conform to those expected values within the app.

For example, if the code says: `sdk.get_number('isEnabled')`
and the payload has: `{ features: [ {key: 'isEnabled', value: true } ] }` then what?
its totally possible for a user to fix the payload, changing it from `true` to `1.0`
it's also possible for there to be two app versions out there (iOS, etc) that call `get_boolean('isEnabled')` and also `get_number('isEnabled')`

Reply:
If you have two different app versions with two different type expectations it seems like you have the opportunity and the motive to add a new feature flag which supports the type you want to use.

IMO, the better behavior is to fail and return the default in the event of a type mismatch.  Because while yes you can coerce some values to some types, in practice I think this leads to a lot of work for SDK engineers, a lot of opportunities for mismatch between SDKs (and SDK versions), and, when its working correctly, a lot of undefined or magic behavior that customer's will guess at rather than know for certain.

If nothing else we should be strict to start and loosen our requirements later when there is a clear demand for it.

> current/cached values that don't conform to those expected values within the app.

I don't think caching is the primary concern here.  This is a migration problem.

If you want to migrate a type the consumer of the type needs to be updated to handle both the new and old case.  Then the producer is updated to the new case.  Finally the consumer drops the old case.  Coercion fixes this in that you don't need to explicitly handle both cases but at the expense of hidden behavior.

IMO, type migration will be uncommon so optimizing for it at the expense of a simplified, direct behavior in the common case is a mistake.

> its totally possible for a user to fix the payload, changing it from true to 1.0
it's also possible for there to be two app versions out there (iOS, etc) that call get_boolean('isEnabled') and also get_number('isEnabled')

If you have a legacy installation using the wrong type, you recognize its the wrong type, and want to change it then the appropriate solution is to either coerce it in your code base explicitly (e.g. `toFloat(get_boolean(key))` or `toBool(get_number(key))`) or migrate to a new key (e.g. `isEnabled` and `isEnabledFixed`).

---

One final thought on coercion.  If coercion is important then we should provide an untyped interface `sdk.get_value(key)` which will return anything.  The user can explicitly or the programming language implicitly can coerce the result however they like depending on the context.  The SDK should remove itself from that situation.